File: backend/app/schemas/knowledge.py
"""
知识库相关Schema
"""
from typing import Optional, List
from pydantic import BaseModel, Field


class KnowledgeBaseCreate(BaseModel):
    """创建知识库请求"""
    name: str = Field(..., description="知识库名称", min_length=1, max_length=100)
    description: Optional[str] = Field(None, description="知识库描述")
    embeddingModel: str = Field(default="bge-m3", description="嵌入模型名称")
    chunkSize: int = Field(default=500, description="文本切片大小", ge=100, le=2000)
    chunkOverlap: int = Field(default=100, description="切片重叠长度", ge=0, le=500)


class KnowledgeBaseUpdate(BaseModel):
    """更新知识库请求"""
    name: Optional[str] = Field(None, description="知识库名称", min_length=1, max_length=100)
    description: Optional[str] = Field(None, description="知识库描述")
    embeddingModel: Optional[str] = Field(None, description="嵌入模型名称")
    chunkSize: Optional[int] = Field(None, description="文本切片大小", ge=100, le=2000)
    chunkOverlap: Optional[int] = Field(None, description="切片重叠长度", ge=0, le=500)
    status: Optional[str] = Field(None, description="状态")


class KnowledgeBaseResponse(BaseModel):
    """知识库响应"""
    id: int
    name: str
    description: Optional[str]
    embeddingModel: str
    chunkSize: int
    chunkOverlap: int
    status: str
    documentCount: int = 0
    createdAt: Optional[str]
    updatedAt: Optional[str]
    createdBy: Optional[int]


class DocumentUpload(BaseModel):
    """文档上传请求"""
    knowledgeBaseId: int = Field(..., description="知识库ID")
    fileName: str = Field(..., description="文件名")
    fileType: str = Field(..., description="文件类型")


class DocumentResponse(BaseModel):
    """文档响应"""
    id: int
    knowledgeBaseId: int
    fileName: str
    filePath: str
    fileType: str
    fileSize: Optional[int]
    pageCount: Optional[int]
    status: str
    errorMessage: Optional[str]
    segmentCount: int
    createdAt: Optional[str]
    updatedAt: Optional[str]


class DocumentSegmentResponse(BaseModel):
    """文档切片响应"""
    id: int
    documentId: int
    knowledgeBaseId: int
    content: str
    segmentIndex: int
    startChar: Optional[int]
    endChar: Optional[int]
    metadata: dict
    createdAt: Optional[str]


class KnowledgeQuery(BaseModel):
    """知识检索请求"""
    knowledgeBaseId: int = Field(..., description="知识库ID")
    question: str = Field(..., description="查询问题", min_length=1)
    topK: int = Field(default=5, description="返回结果数量", ge=1, le=20)


class KnowledgeQueryResult(BaseModel):
    """知识检索结果"""
    segmentId: int
    documentId: int
    documentName: str
    content: str
    score: float
    metadata: dict


class KnowledgeAnswerResponse(BaseModel):
    """知识问答响应"""
    answer: str
    references: List[KnowledgeQueryResult]
    queryTime: float

File: frontend/src/api/knowledge.ts
import request from './index'

export interface KnowledgeBase {
  id: number
  name: string
  description?: string
  embeddingModel: string
  chunkSize: number
  chunkOverlap: number
  status: string
  documentCount?: number
  createdAt: string
  updatedAt: string
  createdBy?: number
}

export interface Document {
  id: number
  knowledgeBaseId: number
  fileName: string
  filePath: string
  fileType: string
  fileSize?: number
  pageCount?: number
  status: string
  errorMessage?: string
  segmentCount: number
  createdAt: string
  updatedAt: string
}

export interface KnowledgeQuery {
  knowledgeBaseId: number
  question: string
  topK?: number
}

export interface KnowledgeQueryResult {
  segmentId: number
  documentId: number
  documentName: string
  content: string
  score: number
  metadata: any
}

export interface KnowledgeAnswer {
  answer: string
  references: KnowledgeQueryResult[]
  queryTime: number
}

export function getKnowledgeBases(params?: { skip?: number; limit?: number }) {
  return request.get<{ list: KnowledgeBase[]; total: number }>('/knowledge-bases', { params })
}

export function getKnowledgeBase(id: number) {
  return request.get<KnowledgeBase>(`/knowledge-bases/${id}`)
}

export function createKnowledgeBase(data: any) {
  return request.post<KnowledgeBase>('/knowledge-bases', data)
}

export function updateKnowledgeBase(id: number, data: any) {
  return request.put<KnowledgeBase>(`/knowledge-bases/${id}`, data)
}

export function deleteKnowledgeBase(id: number) {
  return request.delete(`/knowledge-bases/${id}`)
}

export function getDocuments(kbId: number, params?: { skip?: number; limit?: number }) {
  return request.get<{ list: Document[]; total: number }>(`/knowledge-bases/${kbId}/documents`, { params })
}

export function uploadDocument(kbId: number, file: File) {
  const formData = new FormData()
  formData.append('file', file)
  return request.post<Document>(`/knowledge-bases/${kbId}/documents`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
}

export function deleteDocument(kbId: number, docId: number) {
  return request.delete(`/knowledge-bases/${kbId}/documents/${docId}`)
}

export interface DocumentDetail extends Document {
  totalChars: number
  avgChunkChars: number
}

export interface DocumentSegment {
  id: number
  documentId: number
  knowledgeBaseId: number
  content: string
  segmentIndex: number
  startChar?: number
  endChar?: number
  metadata: any
  charCount: number
  highlight?: string
  createdAt: string
}

export function getDocumentDetail(kbId: number, docId: number) {
  return request.get<DocumentDetail>(`/knowledge-bases/${kbId}/documents/${docId}`)
}

export function getDocumentSegments(
  kbId: number,
  docId: number,
  params?: { skip?: number; limit?: number; keyword?: string }
) {
  return request.get<{ total: number; segments: DocumentSegment[] }>(
    `/knowledge-bases/${kbId}/documents/${docId}/segments`,
    { params }
  )
}

export function buildIndex(kbId: number) {
  return request.post<{ indexedDocuments: number }>(`/knowledge-bases/${kbId}/build-index`)
}

export function queryKnowledge(kbId: number, data: KnowledgeQuery) {
  return request.post<KnowledgeAnswer>(`/knowledge-bases/${kbId}/query`, data)
}
